### ISSUE TITLE
Cython + Pyinstaller

### ISSUE THREAD

**Issue by topiwala** [pyinstaller/pyinstaller#1353]
I have a multi-package multi-directory Cython repository that needs to be packaged. Entry point is a python file. Structure is like this:
```
main_proj
    |
    |-----shell.py (entry point)
    |-----A
            |----__init__.py
            |----foo.so (etc)
   |-------B
            |----__init__.py
            |---foo.so (etc)
            |-----C
                    |---__init__.py
                    |---foo2.so (etc)
```
Now, I can run this directly using the python interpreter by doing python main_proj/shell.py after setting my PYTHONPATH to be the top level directory. The individual so files import from the top

```import main_proj.B.foo (etc)```

pyinstaller fails to find any of the imported modules, giving them as --hidden-imports doesn't seem to work either, as it can't find them even with -p path specified.

Few questions
(a) Is there a multi-package cython example that I could follow?
(b) Should I be writing a spec file, or is pyinstaller smart enough to handle this kind of a directory structure?
(c) Any general direction of how I should be approaching this would be __very__ helpful.

Thanks!

**Comment by topiwala** [comment]
So, C is under B, didn't come out correctly in the pixel art

**Comment by codewarrior0** [comment]
Enclose your pixel art in triple backquotes to preserve formatting (see [Syntax Highlighting](https://help.github.com/articles/github-flavored-markdown/)).

I answered this on the mailing list, but it doesn't look like your problem was what I thought it was. Does `main_proj` have an `__init__.py` file too? (In other words, is it a package module?) If so, are you passing the folder containing `main_proj` as the `-p` argument to PyInstaller? If not, you shouldn't expect `import main_proj.B.foo` to work in Python.

**Comment by topiwala** [comment]
It does contain an __init__.py. 

Packaging the original python repository works fine (without any -p args supplied), but does not work with the Cythonized version.

**Comment by codewarrior0** [comment]
Oh, I almost forgot:

Python version, PyInstaller version, and OS version?

**Comment by topiwala** [comment]
Python: 2.7.6
pyinstaller: 2.1
OS: Ubuntu 14.04.1 LTS (64-bit)
Kernel version: 3.13.0-32-generic

**Comment by topiwala** [comment]
If it helps, I noticed that in Analysis.TOC placed under build, some of the
files aren't even present in the main directory. The Analysis.TOC contains
the tuple, when it should be

('A.B.C.constants', '/home/topi/WORK/A/B/C/constants.pyc', 'PYMODULE')


constants.so instead, since constants.pyc is not present.


Thanks

**Comment by codewarrior0** [comment]
It *should* be finding those .so files without any special action from you.

The first thing I would try to find out is whether it is including those .so files in the wrong place, or simply not including them. Do a `find -iname foo.so` on the output folder under `dist/` and see if it turns up.

> ('A.B.C.constants', '/home/topi/WORK/A/B/C/constants.pyc', 'PYMODULE')

Actually, never mind! Are you sure constants.pyc is not present? Try removing all .pyc files from your project. I remember noticing something that might be a bug, where it prefers the .pyc file if it finds multiple modules with the same name in the same folder.

**Comment by topiwala** [comment]
Yes, constants.pyc is not present, and constants.so is. The only .pyc files
present (and these are needed for the cross directory imports) are the
__init__.py[c] which allow the so files in this directory to be importable.

I do see some so files under dist, but not nearly as many as the number I'm
expecting, and there seems to be no way to tell why it picked these up.

Is there any debugging flag that I could use to see the progress as it goes
through stitching together the dependencies?

**Comment by codewarrior0** [comment]
`--log-level=DEBUG` will turn on the most verbose logging, but it doesn't trace the module finding down to the individual modules. I think the next thing you should try is `--clean`, because the .pyc files it's finding might be in the `build` folder.

**Comment by codewarrior0** [comment]
BTW, If you're going to look at the TOC files some more, there's a `pyi_pprint-toc` command for pretty-printing them.

**Comment by codewarrior0** [comment]
[Further discussion on the cython-users group.](https://groups.google.com/forum/#!topic/cython-users/n-fbsOwykrA)

**Comment by topiwala** [comment]
I'm still not sure what files go where in the Analysis step. Here's a small snippet of code:
pkgObj.py_files is a list of __init__.py files
pkgObj.ext_files is a list of so files, which are basically my python sources in the main project that have been cythonified (and so are really like python modules) that have the 'EXTENSION' type
pkgObj.data_files are data_files, where name = path.

It generates the executable file, but fails with import errors on the so files. Where do the ext_files get added, to a.pure, or to a.binary? As what type, 'EXTENSION", or 'PYMODULE'?

Am I missing something else here?

a = Analysis([os.path.join(package_name, entry_point)],
              pathex=['.'],
              hiddenimports=[],
              hookspath=None)

for lst in pkgObj.py_files:
    a.pure += TOC([tuple(lst)])
for lst in pkgObj.ext_files:
    a.pure += TOC([tuple(lst)])
for lst in pkgObj.data_files:
    a.datas += TOC([tuple(lst)])

pyz = PYZ(a.pure)
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          name=package_name,
          debug=False,
          strip=None,
          upx=False,
          console=True)

**Comment by codewarrior0** [comment]
I edited your comment with triple backquotes to improve formatting.

ext_files should go into `a.binaries` as `EXTENSION`.

**Comment by topiwala** [comment]
I've tried all the different options.
To the a.pure tuple, I've added tuples of this form for all the directory paths in main_proj
[main_proj.A.B.C, /home/topi/main_proj/A/B/C/__init__.py, "PYMODULE"]

To the a.binaries set of tuples, I've added tuples of the form
[main_proj.A.B.C.d, /home/topi/main_proj/A/B/C/d.so, "EXTENSION"]

To the a.datas set of tuples, I've added tuples of the form
[main_proj/A/B/C/e.cpp, /home/topi/A/B/C/e.cpp, "DATA"]

I've looked at the resulting executable dist/main_proj using pyi-archive_viewer, and the number of so files from main_proj in the archive correspond with the number of so files in my main_proj repo.

While running the executable, I still get the following error:

ImportError: No module named xx

where xx is a cython module.

**Comment by codewarrior0** [comment]
Hum, that's a bit mystifying. For what it's worth, I successfully use cython with pyinstaller in my own project, but the few cython files I have are all directly imported from python files and those python files that *are* imported from cython are also imported elsewhere, so everything gets included.

Could you try adding a bunch of "dummy" imports of all of your cython files into your main script (or a file imported by it)? Just to make sure the modulegraph in Analysis picks up on them, so you can see if they can get included normally.

**Comment by topiwala** [comment]
The problem was staring at me. I erroneously passed the first argument to EXE as the pyz archive, when the first should be the scripts tuple.

Thankfully, now I'm getting ImportErrors in a completely different module (not part of my project), pyi_os_path. More debug!

**Comment by topiwala** [comment]
Hmm...looking at the more detailed logs from the -v option, it appears that the argument order matters. And it seems like I get further along in processing the executable with pyz being the first argument, and don't hit the import pyi_os_path error (which seems to be creating a bootloader environment for execution of my executable), but hit the import error of my cython module later....

**Comment by htgoebel** [comment]
There should be no need of adding files the way you do in https://github.com/pyinstaller/pyinstaller/issues/1353#issuecomment-126155117, hidden imports should suffice.

If adding the module as hidden imports does not work, please submit a *minimal* example so we can debug it. You may take https://gist.github.com/ronen/024cdae9ff2d50488438 as an example.

**Comment by topiwala** [comment]
Does this TOC added to anal.binaries imply that the relative path

main_proj/A/B/C will be created during packaging, and d.so will be placed there? 

[main_proj.A.B.C.d, /home/topi/main_proj/A/B/C/d.so, "EXTENSION"]

It looks like d.so is COPIED from the build directory (argument 2) specified in the tuple, and placed in the top level dist directory where it's named main_proj.A.B.C.d

That's not what I want, since this file is actually imported inside other python files in the repo (like so)

import main_proj.A.B.C.d

**Comment by codewarrior0** [comment]
That's as expected. I'm not sure why this is, but PyInstaller uses a custom importer (`CExtensionImporter`) for extensions that uses the dotted module name as the filename instead of looking in subfolders.

**Comment by codewarrior0** [comment]
Maybe it would help to enable Python's own verbose logging, which traces the import machinery and search paths. Add the '-v' option to the TOCs passed to EXE like so:

```
exe = EXE(pyz,
          a.scripts,
          a.binaries,
          a.zipfiles,
          a.datas,
          [('-v', '', 'OPTION')],  # <<<<<
          name=package_name,
          debug=False,
          strip=None,
          upx=False,
          console=True)
```

**Comment by topiwala** [comment]
Actually, I have managed to get past my initial import issues, and exact imports (with fully qualified path names) work fine. Breaking it out into exe and collect helped. I did have to add to a.pure (PYMODULE) & a.binary (EXTENSION) all the modules that Analysis was unable to find.

However, since this is a multi-package, multi-directory python repo, I have quite a few places that use sophisticated python imports (relative imports, putting imports in __init__.py), and these are the ones that are failing now. Is there ANY way around this? 

Would  a post-processing script that creates these directories, and puts in the symlinks to the top level so files actually work? How would one get around imports that are directly coded into __init__.py in the directory?

**Comment by codewarrior0** [comment]
> How would one get around imports that are directly coded into init.py in the directory?

This is why you should try to use hiddenimports instead of inserting modules into the TOCs. hiddenimports will scan those modules for further imports and include them too.

**Comment by htgoebel** [comment]
> This is why you should try to use hiddenimports instead of inserting modules into the TOCs.

ACK! @topiwala It looks like you are doing much too much per hand which PyInstaller does automatically. Throw away you .spec-file, open your mind and start from scratch.

**Comment by linkmauve** [comment]
@topiwala, I just compiled my Cython project using PyInstaller, all I had to do for it to detect my Cython modules was to list all of them in import statements in the scriptname passed to pyinstaller.

I used something similar to that:
```
if False:
    import package.module1
    import package.module2
    import package.subpackage.module3
```

**Comment by topiwala** [comment]
htgoebel:
hidden imports worked, and I have moved to that. I had to fix my tree to not perform relative imports, and that helped as well. There were some special libraries not found by default, so I had to coerce them too by adding them to the hidden imports. These are 
https://docs.python.org/2/library/code.html
https://docs.python.org/2/library/cgi.html
https://docs.python.org/2/library/bisect.html

However, I still can't seem to get the data files into the package correctly. 
I'm packaging them as

for lst in pkgObj.data_files:
    a.datas += TOC([tuple(lst)])
where each lst is

[relative_path, full_path, 'DATA']

Is there something else also I need to do? I'm trying to create a one file executable.
